### PR TITLE
Add support for generic probes on Pad.

### LINKED ISSUE #189: Add generic probe support to Pad
Add `addProbe()` method to `Pad` to better follow upstream API.

Map PadProbeType and PadProbeInfo, with support for extracting events, buffers, etc.

Possibly deprecate existing probe-based methods.

### PR DESCRIPTION
**Pull request by neilcsmith-net**
Add support for generic probes on Pad, better matching upstream API. Fixes #189

### CODE REVIEW

File: test/org/freedesktop/gstreamer/PadProbeTypeTest.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+/**
+ *
+ */
```
Comment (woglinde) [inline test/org/freedesktop/gstreamer/PadProbeTypeTest.java:34]: Do you want leave this in public code?

File: test/org/freedesktop/gstreamer/PadProbeTypeTest.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+public class PadProbeTypeTest {
+    
+
+    @Test
+    public void testCombinations() throws Exception {
+        for (Field field : PadProbeType.class.getFields()) {
+            if (Set.class.isAssignableFrom(field.getType())) {
+                Set<PadProbeType> flags = (Set<PadProbeType>) field.get(null);
+//                System.out.println(field.getName() + " : " + flags);
```
Comment (woglinde) [inline test/org/freedesktop/gstreamer/PadProbeTypeTest.java:43]: Consider using a logger here or remove it?

File: test/org/freedesktop/gstreamer/PadProbeTypeTest.java (hunk 1 above)
Comment (neilcsmith-net) [reply test/org/freedesktop/gstreamer/PadProbeTypeTest.java:34]: Only excuse is it's only in a test file! :smile:

File: test/org/freedesktop/gstreamer/PadProbeTypeTest.java (hunk 2 above)
Comment (neilcsmith-net) [reply test/org/freedesktop/gstreamer/PadProbeTypeTest.java:43]: Right now, no, but in general we need to refactor a bunch of stuff in tests.  There's a whole mix of stuff that's output to System.out, some commented out, some not.  It's useful stuff at times locally.  Ideally will refactor the lot and put behind a flag / through logging.

**Comment by neilcsmith-net** [comment]
Need to check / add in support for probe removal return value - must release Pad.PROBE implementation to allow it to be GC'd.  Add tests for this too.  Storing probes and IDs in a Map<PROBE, NativeLong> inside Pad might be a better idea than current behaviour.

File: test/org/freedesktop/gstreamer/PadProbeTypeTest.java (hunk 2 above)
Comment (lafoletc) [reply test/org/freedesktop/gstreamer/PadProbeTypeTest.java:36]: Need tests on Query

**Comment by neilcsmith-net** [comment]
@lafoletc I agree needs a Query test. However, porting allocation query test didn't seem to work correctly.  I also need to look at fixing up ProbeTester, but only thing I seemed to get was a Caps query with that pipeline when running a query probe IIRC.  Feel free to take a look if you want.

**Pull request merged**